kroxylicious/src/test/java/io/kroxylicious/proxy/internal/InternalCompletionStageTest.java

```diff
@@ -0,0 +1,308 @@
+
+                Arguments.of("handleAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.handleAsync((u, t) -> tallyInvocations(), e)),
+                Arguments.of("whenCompleteAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.whenCompleteAsync((u, t) -> tallyInvocations(), e)),
+
+                Arguments.of("exceptionallyAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.exceptionallyAsync((t) -> tallyInvocations(), e)),
+                Arguments.of("exceptionallyComposeAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s
+                                .exceptionallyComposeAsync((t) -> CompletableFuture.completedStage(tallyInvocations()), e)),
+
+                Arguments.of("acceptEitherAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.acceptEitherAsync(other, (u) -> tallyInvocations(), e)),
+                Arguments.of("applyToEitherAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.applyToEitherAsync(other, (u) -> tallyInvocations(), e)),
+
+                Arguments.of("thenAcceptBothAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.thenAcceptBothAsync(other, (u1, u2) -> tallyInvocations(), e)),
+                Arguments.of("thenCombineAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.thenCombineAsync(other, (u1, u2) -> tallyInvocations(), e)),
+
+                Arguments.of("runAfterBothAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.runAfterBothAsync(other, () -> CompletableFuture.completedStage(
+                                tallyInvocations()), e)),
+                Arguments.of("runAfterEitherAsync",
+                        (BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>>) (e, s) -> s.runAfterEitherAsync(other,
+                                () -> CompletableFuture.completedStage(
+                                        tallyInvocations()),
+                                e)));
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource
+    void executorFormAsyncMethodsUsesCallerExecutor(String name, BiFunction<Executor, CompletionStage<Void>, CompletionStage<Void>> func) throws Exception {
+
+        var callerExecutor = Executors.newSingleThreadExecutor(TallyThread::new);
+        try {
+            var callerExecutorThread = ((TallyThread) callerExecutor.submit(Thread::currentThread).get());
```

Comment:
Would it be cleaner to have the `TallyThread` increment an injected `AtomicInteger` which the tests could then assert the counts on rather than submitting and casting the result?

Reply:
I've done away with the TallyThread and used the TPE#getCompletedTaskCount instead.